### ISSUE TITLE
Replace pickle with json everywhere in Leo

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#800]
[Don't Pickle Your Data](http://www.benfrederickson.com/dont-pickle-your-data/) recommends using [python's json module](https://docs.python.org/2/library/json.html) instead of pickle.

Terry suggests that Leo's use of pickle can be fixed easily.  If so, there would be no reason to use json.

**Comment by tbnorth** [comment]
I'm not sure how important / easy / bang for buck ish this is.

JSON can't:

 - represent sets
 - represent datetimes
 - represent any Python object that doesn't provide JSON serialization callbacks
 - *use anything but strings as keys*
 - represent NaN, +/-Inf etc.

As long as none of those things turn out to be important, switching out might be worth while.  It's possible that the breakage of settings between 2 and 3 could be fixed by telling the Py3 code to use a lower Pickle format version, which would be almost no work at all.

I guess I don't have a clear view of the size of the problem vs. the work needed for the solution.

**Comment by edreamleo** [comment]
On Sun, Mar 11, 2018 at 9:07 PM, Terry Brown <notifications@github.com>
wrote:

> JSON can't:
>
>    - represent sets
>    - represent datetimes
>    - represent any Python object that doesn't provide JSON serialization
>    callbacks
>    - *use anything but strings as keys*
>    - represent NaN, +/-Inf etc.
>
> ​Phooey.  I didn't know that.​


> It's possible that the breakage of settings between 2 and 3 could be fixed
> by telling the Py3 code to use a lower Pickle format version, which would
> be almost no work at all.
>
​Let's try that.  There shouldn't be any security implications in caching
data with pickle.

Edward

**Comment by edreamleo** [comment]
David Szent-Györgyi: I recall a blog posting about the limitations of JSON that are overcome by AXON. The posting states that unlike JSON, AXON supports comments, provides syntax for decimal and date/time values, and supports streams of values. 

The Pypi package pyaxon runs on Python 2.7, 3.3, 3.4, 3.5. Source code is available on GitHub. 

In case the original blog posting is not available, it is quoted extensively in a tuicool posting.

**Comment by tbnorth** [comment]
I think on the "switching between 3 and 2" issue we're just out of luck.
https://github.com/leo-editor/leo-editor/blob/devel/leo/core/leoCache.py#L503
Because pickle needs to map back to class definitions in .py(c) files that need to exist at de-pickling time, and because 2.x -> 3.x is a breaking change as far as those mappings are concerned, I just don't think Python 3 pickles are readable in Python 2.  I spent a long time on this back when we first supported 3 and datetime.datetime pickles in todo.py were a problem.  I ended up migrating to plain text dates.

Personally I could run 3 everywhere, but like to run 2 when editing codebases written in 2 so I don't get spurious error reports for things written in 2 interpreted in 3.  But more generally I don't think we want to  force abandoning 2 just because of this issue.

Ultimately pickled info.'s not that important - caches can be rebuilt and other uses are minimal, e.g. bookmarks.py just stores the UNL of @bookmarks node, free_layout does store layouts and the name of the current layout there. I'll probably target JSON for the qdock layout persistence.

So if you want to switch from 3 to 2, delete the cache and recreate what's lost, and don't make it part of your workflow.
